### PR TITLE
[TPU] Support collective communications in XLA devices

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR adds support for collective communications in XLA devices (TPU). It is simply implemented by falling back to `xm.all_reduce` and `xm.all_gather` for TPU devices. One difference is the gather operation in logits processor, where `gather` is replaced by `all-gather` to meet the SPMD restriction in XLA.

### CODE REVIEW

**Comment by WoosukKwon** [comment]
This PR was part of #5871, but separated out to get a quick review.

File: vllm/platforms/interface.py [hunk 1]
```diff
@@ -23,6 +23,9 @@ def is_rocm(self) -> bool:
     def is_tpu(self) -> bool:
         return self._enum == PlatformEnum.TPU
 
+    def is_xla(self) -> bool:
+        return self.is_tpu()
+
```
Comment (youkaichao) [inline vllm/platforms/interface.py:28]: why do we need this? isn't `is_tpu()` enough?

File: vllm/distributed/parallel_state.py [hunk 2]
```diff
@@ -125,6 +129,7 @@ class GroupCoordinator:
     pynccl_comm: Optional[Any]  # PyNccl communicator
     ca_comm: Optional[Any]  # Custom allreduce communicator
     mq_broadcaster: Optional[Any]  # shared memory broadcaster
+    use_xla: bool  # Whether to use PyTorch XLA communicator
```
Comment (youkaichao) [inline vllm/distributed/parallel_state.py:132]: does tpu platform support NCCL? if not, creating these communicators might lead to error.

Comment (WoosukKwon) [reply vllm/distributed/parallel_state.py:132]: TPU doesn't support NCCL, but I didn't see any error with the other communicators.

Comment (WoosukKwon) [reply vllm/distributed/parallel_state.py:132]: The TPU backend uses gloo backend in addition to the distributed runtime in `xm`. Maybe that's the reason.

File: vllm/platforms/interface.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/platforms/interface.py:28]: You're right. Neuron actually uses XLA, but it does not use `vllm/distributed`.

Comment (WoosukKwon) [reply vllm/platforms/interface.py]: Fixed!

File: vllm/distributed/parallel_state.py (hunk 2 above)
Comment (youkaichao) [inline vllm/distributed/parallel_state.py:132]: `use_xxx` is a initialization parameter, and we usually hold `communicator` inside group coordinator.

Can you add a `tpu_communicator` under https://github.com/vllm-project/vllm/tree/main/vllm/distributed/device_communicators ?

One additional benefit, is that you can implement the `gather` logic to `allgather`, without intrusive change to `logits_processor.py` .

File: vllm/model_executor/layers/logits_processor.py [hunk 3]
```diff
@@ -76,7 +80,15 @@ def _get_logits(self, hidden_states: torch.Tensor,
...
-        logits = tensor_model_parallel_gather(logits)
+        if self.use_gather:
+            logits = tensor_model_parallel_gather(logits)
+        else:
+            # Gather is not supported for some devices such as TPUs.
+            # Use all-gather instead.
+            # NOTE(woosuk): Here, the outputs of every device should not be None
+            # because XLA requires strict SPMD among all devices. Every device
+            # should execute the same operations after gathering the logits.
+            logits = tensor_model_parallel_all_gather(logits)
```
Comment (youkaichao) [inline vllm/model_executor/layers/logits_processor.py:91]: ideally, this dispatching logic should not exist in logits processors, but handled by TPU communicators.

File: vllm/distributed/parallel_state.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/distributed/parallel_state.py:132]: > One additional benefit, is that you can implement the gather logic to allgather, without intrusive change to logits_processor.py .

This is actually not the case because the TPU backend explicitly requires all-gather, which means each device's output should not be None. If we implement gather by using `all-gather` and outputting None for non-root ranks, XLA will raise an error.

File: vllm/model_executor/layers/logits_processor.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/logits_processor.py:91]: Again, this is hard because "gather" means outputting None in non-root devices.

Comment (youkaichao) [reply vllm/model_executor/layers/logits_processor.py:91]: sorry i don't understand. per my understanding, you can use `allgather`, and then discard the output in non-root devices.

Comment (WoosukKwon) [reply vllm/model_executor/layers/logits_processor.py:91]: As written in the comment, XLA requires strict SPMD. The graph executed by the devices should all be the exactly same. If the logits is None in non-root devices, they will have a graph that does not execute the following lines of code such as `logits = logits[:, :self.org_vocab_size]` and sampler. This will cause the difference between the graph for the root device, which XLA does not allow.

File: vllm/distributed/device_communicators/tpu_communicator.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        local_rank: int,
+        world_size: int,
+    ):
+        del group  # Unused.
+        if not current_platform.is_tpu():
+            self.disabled = True
+            return
+        self.disabled = False
+
+        pjrt.initialize_multiprocess(local_rank, world_size)
```
Comment (youkaichao) [inline vllm/distributed/device_communicators/tpu_communicator.py:27]: you can get rank and world size from the `group`

Comment (WoosukKwon) [reply vllm/distributed/device_communicators/tpu_communicator.py:22]: Thanks for letting me know! Updated the PR. PTAL.

Comment (youkaichao) [reply vllm/distributed/device_communicators/tpu_communicator.py:22]: do you need to remove the code inside tpu worker? I don't know if pjrt and xm support initialization for multiple times.

File: vllm/lora/layers.py [hunk 5]
```diff
@@ -1067,6 +1067,10 @@ def scale(self):
     def soft_cap(self):
         return self.base_layer.soft_cap
 
+    @property
+    def use_gather(self):
+        return self.base_layer.use_gather
+
```
Comment (youkaichao) [inline vllm/lora/layers.py:1073]: what is this used for?

File: vllm/distributed/device_communicators/tpu_communicator.py (hunk 4 above)
Comment (WoosukKwon) [reply vllm/distributed/device_communicators/tpu_communicator.py:22]: @youkaichao Good point. That's actually updated in #5871. In the current main branch, there's no code initializing the XLA's distributed runtime.

File: vllm/lora/layers.py (hunk 5 above)
Comment (WoosukKwon) [reply vllm/lora/layers.py:1073]: This class wraps the logits processor class and requires exposing every property under `self` this way. Without this, we'll see failures in lora tests.

**Review by youkaichao — APPROVED** [review]
Thanks for addressing my comments!

**Comment by WoosukKwon** [comment]
@youkaichao Thanks for your review and suggestion!
